### PR TITLE
adds "nosource" modifier and parameter for ns/class/apl/lib directives in DBuild (#18)

### PR DESCRIPTION
**Pull request by mbaas2**
...to avoid adding "source as typed" to the workspace which could cause issues when using the same code on classic AND unicode interpreters.

### CODE REVIEW

File: DyalogBuild.dyalog [hunk 1]
```diff
@@ -312,10 +316,13 @@
...
-                              res←0({0::⍵ ⋄ ⍎⍵}target)0 ⎕SE.SALT.Load fl,' ',options
-     
-                              ⍝source←⎕SE.SALT.Load fl,' -source=no'  ⍝ unfortunately need two calls to establish function & get source
-                              ⍝source←1⊃⎕NGET fl
+                              :If {(1=≢⍵)∧⊃1↑,⍵}~(⎕NC⍕target)∊0 9  ⍝ deal with name clashes
+                              :AndIf (,'#')≢,⍕target
+                                  ('target="',(⍕target),'" exists already with ⎕NC=',(⍕⎕NC target),' and is protected')⎕SIGNAL(∨/' -protect'⍷options)/11
+                                  ⎕EX target
+                              :EndIf
+                              ref←{0::⍵ ⋄ ⍎⍵}target
```
Comment (bpbecker) [inline DyalogBuild.dyalog:324]: This seems a little presumptuous - on any error return the argument?

**Review by bpbecker — COMMENTED** [review]
- This PR has more than one purpose - e.g. Adding support for nosource AND upgrading code coverage.
- From an OCD standpoint, the mixed-case ⎕names bothers me - there's even a line that has both ⎕NC and ⎕nc in it.
- my comments and review are "tactical" and address questions at a very low level.  To evaluate it at a strategic design level requires more bandwidth than I have available at present.
- I assume that the changes have been tested on multiple platforms with various versions of Dyalog APL.

File: DyalogBuild.dyalog [hunk 2]
```diff
@@ -819,7 +827,7 @@
           setups←' 'Split args.setup
       :EndIf
      
-      r←''
+      r←''  ⍝ must be global here, is localized in the calling fn
```
[the remark this answers is not in the document]
Comment (mbaas2) [reply DyalogBuild.dyalog:830]: r is the result of the calling fn - and XTest ist only called by that fn. You made suggestions re. the use of ⎕TSYNC to handle this more elegantly, but I felt that was  too instrusive to address with this update.

File: Tests/CompCheck/Run.aplf [hunk 3]
```diff
@@ -36,15 +43,15 @@
...
+     s←('∧.*⍝ CompCheck: ignore.*$'⎕R'')s
      s←{1↓¨(⍵=⎕UCS 10)⊂⍵}(⎕UCS 10),s  ⍝ nested lines
      nq←{0=≠\⍵=''''}¨s                ⍝ not quoted (leaves one quote, but that is no problem for this purpose)
      uncmnt←nq{~(∨\⍺∧⍵='⍝')}¨s        ⍝ not commented
      s←∊((nq∧uncmnt)/¨s),¨⎕UCS 10     ⍝ back into simple string
 
      :For i :In ⍸(1↓Features∆)[;1]>minVers
          c←i⊃features
-         :If 0<≢l←(('(?:\W|^)',(2⊃c),'(?:\W|$)')⎕S'%'⍠'Mode' 'L')s
+         :If 0<≢l←(('(?:\W|∧)',(2⊃c),'(?:\W|$)')⎕S'%'⎕U2360'Mode' 'L')s
```
[the remark this answers is not in the document]
Comment (mbaas2) [reply Tests/CompCheck/Run.aplf]: Updated, thanks! ( https://github.com/Dyalog/DBuildTest/commit/eb3d5f3c58e02d1b6cd7de6744a5fdc401904310 )

File: DyalogBuild.dyalog [hunk 4]
```diff
@@ -2007,6 +2041,10 @@
               r,←'' 'Optional modifiers are:'
               r,←⊂'    -clear[=NCs]    expunge all objects, optionally of specified nameclasses only'
               r,←⊂'    -halt           halt on error rather than log and continue'
+              :If 19≤DyaVersion
+                  r,←⊂'    -nosource       do not preserve "source-as-typed" (neccessary if you want to create workspaces'
```
[the remark this answers is not in the document]
Comment (mbaas2) [reply DyalogBuild.dyalog:2052]: Updated, thanks! ( https://github.com/Dyalog/DBuildTest/commit/e1ddcb801e8ac8a770e1fc42e21d492c9888cf39 )

File: DyalogBuild.dyalog [hunk 5]
```diff
@@ -1682,12 +1711,16 @@
...
+                              :If (⊂2↑icon)∊'./' '.\'
+                                  icon←path,2↓icon
+                              :EndIf
+                              pars←'.' 'Bind'wsid(type)(GetNumParam'flags')(GetParam'resource')(icon)('"'~⍨GetParam'cmdline')(det)
                               command←'2 ⎕NQ ',∊{''≡0↑⍵:'''',⍵,''' ' ⋄ (⍕⍵),' '}¨¯1↓pars
                               command←command,(0<≢det)/' (',(⍕⍴det),'⍴',(∊{''≡0↑⍵:'''',⍵,''' ' ⋄ (⍕⍵),' '}¨det),')'
                               2 #.⎕NQ pars
                           :Else
-                              Log'Builds using "type" (to create something else than a DWS) are only supported on Windows!'
+                              ('Type' 'E')Log'The "type" parameter of TARGET is supported on v19 and better!'
```
[the remark this answers is not in the document]
Comment (mbaas2) [reply DyalogBuild.dyalog]: Updated, thanks!

File: DyalogBuild.dyalog [hunk 6]
```diff
@@ -1594,6 +1608,23 @@
...
+          :EndIf
+          :If (1=GetNumParam'nosource')∧0≠2 args.Switch'nosource'
+          :OrIf 1=2 args.Switch'nosource'
+              {}5171⌶#
+              {}5172⌶0
+          :EndIf
+      :Else
+          :If 0<≢GetParam'nosource'
+          :OrIf nosource>¯1
+              ('Type' 'W')Log'Using "nosource" requires at least Version 19'
```
[the remark this answers is not in the document]
Comment (mbaas2) [reply DyalogBuild.dyalog]: Updated, thanks!

File: DyalogBuild.dyalog [hunk 7]
```diff
@@ -878,7 +886,11 @@
               :If null≢args.coverage ⍝ if switch is set
               :AndIf (1↑1⊃⎕VFI⍕args.coverage)∨1<≢args.coverage  ⍝ and we have either numeric value for switch or a longer string
               :AndIf 0=⎕NC'CoCo'   ⍝ only neccessary if we don't have an instance yet...
-                  home←1⊃⎕NPARTS SALT_Data.SourceFile  ⍝ CompCheck: ignore
+                  :If 9=⎕NC'SALT_Data'
+                      home←1⊃⎕NPARTS SALT_Data.SourceFile  ⍝ CompCheck: ignore
+                  :Else
+                      home←1⊃⎕NPARTS 50 ⎕ATX 1⊃⎕SI
```
[the remark this answers is not in the document]
Comment (mbaas2) [reply DyalogBuild.dyalog:894]: I added an explicit check for that.

File: DyalogBuild.dyalog (hunk 1 above)
Comment (mbaas2) [reply DyalogBuild.dyalog]: Reformulated that statement.

File: DyalogBuild.dyalog (hunk 2 above)
[the remark this answers is not in the document]
Comment (bpbecker) [reply DyalogBuild.dyalog]: Save this for the next iteration...

File: DyalogBuild.dyalog [hunk 8]
```diff
@@ -817,7 +819,7 @@
       :EndIf
       filter←{w←⍵ ⋄ ~∨/'?*'∊⍵: ⍵ ⋄ ((w='?')/w)←'.' ⋄ ((w='*')/w)←⊂'.*' ⋄ ∊⍵}filter
       :If null≢filter
-      :AndIf 0∊⍴fns←(0<≢filter∘{(⍺⎕s'&')⍵}¨fns)/fns
+      :AndIf 0∊⍴fns←⊃,/{(∊0<⍴¨⍵)/⍵}filter∘{(⍺ ⎕S'%')⍵}¨fns
```
Comment (bpbecker) [inline DyalogBuild.dyalog:823]: might be better to write filter ⎕S '%'⊢fns

File: DyalogBuild.dyalog [hunk 9]
```diff
@@ -1687,6 +1689,7 @@
                   :Trap DEBUG↓0 ⍝ yes, all trap have a halt/ after them - this one doesn't and shouldn't.
                       :If ~0∊⍴type←GetParam'type'
                           :If DyaVersion≥19
+                          :orif _isWin∧DyaVersion≥13
```
Comment (bpbecker) [inline DyalogBuild.dyalog]: might be good to add a comment to the effect of saving is available on all platforms from v19.0 and Windows from v13.0

File: DyalogBuild.dyalog [hunk 10]
```diff
@@ -1893,7 +1900,7 @@
           :If type=0  ⍝ only add this information if Log came w/o explicit type
               type←3   ⍝ and the default message type is "Error"
               :If (⎕DR msg)=326
-                  msg←'code returned data with unsupported ⎕DR=326'
+                  ⎕←msg←'code returned data with unsupported ⎕DR=326'
```
Comment (bpbecker) [inline DyalogBuild.dyalog:1909]: remove the ⎕←

**Review by bpbecker — COMMENTED** [review]
A couple more minor tweaks and I think we'll be good!

**Review by bpbecker — APPROVED** [review]
Looks like you've addressed all the items we discussed!

**Review by bpbecker — APPROVED** [review]

**Pull request merged**